morf-core/src/main/java/org/alfasoftware/morf/upgrade/UpgradePathFinder.java

```diff
@@ -214,6 +214,45 @@ private boolean schemasNotMatch(Schema targetSchema, Schema trialSchema, String
     }
   }
 
+  /**
+   * Finds discrepancies between a provided upgrade audit map and the list of steps to apply.
+   * This method compares the UUIDs associated with each upgrade step in the provided list with the corresponding UUIDs
+   * in the upgrade audit map. If discrepancies are found, it logs an error message and throws an IllegalStateException.
+   *
+   * @param upgradeAudit A Map representing the upgrade audit information, where keys are step names and
+   *                     values are corresponding UUIDs.
+   * @throws IllegalStateException If discrepancies are found between upgrade steps and their associated UUIDs in the
+   *                               provided map.
+   */
+  public void findDiscrepancies(Map<String, String> upgradeAudit) {
+
+    List<String> discrepancies = stepsToApply.stream()
+            .filter(step -> isStepDuplicated(step, upgradeAudit))
+            .map(step -> step.getUuid().toString())
+            .collect(Collectors.toList());
+
+    if (!discrepancies.isEmpty()) {
+      String message = String.format("Some upgrade steps could have run before with different UUID. The following have been detected:%n%s",
+              String.join("\n", discrepancies));
+      log.error(message);
+      throw new IllegalStateException(message);
+    }
+  }
+
+  /**
+   * Checks if a given CandidateStep is duplicated based on its name in the upgradeAudit map.
+   * A CandidateStep is considered duplicated if it has the same name but a different UUID
+   * compared to the entries in the upgradeAudit map.
+   *
+   * @param step The CandidateStep to be checked for duplication.
+   * @param upgradeAudit A Map<String, String> representing the upgrade audit information, where keys are step names,
+   *                     and values are corresponding UUIDs.
+   * @return {@code true} if the CandidateStep is duplicated, {@code false} otherwise.
+   */
+  private boolean isStepDuplicated(CandidateStep step, Map<String, String> upgradeAudit) {
+    String name = step.getName();
+    return upgradeAudit.get(name) != null && !upgradeAudit.get(name).equals(step.getUuid().toString());

```

Comment:
You can use `java.util.Objects.equals(Object, Object)` to avoid needing the null check.
The null check actually needs to compute the hash and look into the map, hence better avoiding it.